Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py

```diff
@@ -0,0 +1,110 @@
+"""Read-only and read/write database connection pools definitions"""
+import logging
+import time
+import typing
+from collections import deque
+from contextlib import contextmanager
+
+import psycopg2
+
+from utils.singleton import Singleton
+
+
+class PoolBaseClass:
+
+    """Base class for creating connection pools"""
+
+    def __init__(self,
+                hostname: str,
+                database_name: str,
+                username: str,
+                password: str,
+                pool_size: int=2,
+                ) -> None:
+        """Initialize pool"""
+
+        self.hostname = hostname
+        self.database_name = database_name
+        self.username = username
+        self.password = password
+        self.pool_size = pool_size
+        self._pool = deque()
+
+    def create_connection(self) -> psycopg2.connect:
+        """Create connection and add it to the pool"""
+
+        try:
+            connection =  psycopg2.connect(
+                dbname=self.database_name,
+                host=self.hostname,
+                user=self.username,
+                password=self.password,
+                )
+            logging.info('Connection successful!')
+            self.pool_size -= 1
+            return connection
+        except psycopg2.OperationalError as exception:
+            logging.critical('Connection failed! \
+                See exception message and try again!')
+            raise psycopg2.OperationalError(exception)
+
+    def get_connection(self) -> psycopg2.connect:
+        """Get connection from the pool"""
+
+        # Create and return connection
+        # If the count is less than the pool max size
+        if self.pool_size > 0:
```

Comment:
What about multiprocess? 
Please try to add `lock`. Check my recommendation with lock :)

Reply:
You need to merge your 58 line + 63-66 + 69-71 lines :). Your code has issue with multithreading. Lock is code that allows only access for 1 thread at 1 time. It means  `lock` should be in the same loop as all other checks. Please check my recommendation about  this.

Reply:
If I understand you correctly, I should wrap the whole method in lock?